controller.go

```diff
@@ -107,9 +118,72 @@ func New() (NetworkController, error) {
 	if err := initDrivers(c); err != nil {
 		return nil, err
 	}
+
+	if err := c.initConfig(configFile); err == nil {
+		if err := c.initDataStore(); err != nil {
+			// Failing to initalize datastore is a bad situation to be in.
+			// But it cannot fail creating the Controller
+			log.Warnf("Failed to Initialize Datastore : %v", err)
+		}
+		if err := c.initDiscovery(); err != nil {
+			// Failing to initalize discovery is a bad situation to be in.
+			// But it cannot fail creating the Controller
+			log.Warnf("Failed to Initialize Discovery : %v", err)
+		}
+	} else {
+		// Missing Configuration file is not a failure scenario
+		// But without that, datastore cannot be initialized.
+		log.Debugf("Unable to Parse LibNetwork Config file : %v", err)
+	}
+
 	return c, nil
 }
 
+const (
+	cfgFileEnv     = "LIBNETWORK_CFG"
+	defaultCfgFile = "/etc/default/libnetwork.toml"
+)
+
+func (c *controller) initConfig(configFile string) error {
+	cfgFile := configFile
+	if strings.Trim(cfgFile, " ") == "" {
+		cfgFile = os.Getenv(cfgFileEnv)
+		if strings.Trim(cfgFile, " ") == "" {
+			cfgFile = defaultCfgFile
+		}
+	}
+	cfg, err := config.ParseConfig(cfgFile)
+	if err != nil {
+		return ErrInvalidConfigFile(cfgFile)
+	}
+	c.Lock()
+	c.cfg = cfg
+	c.Unlock()
+	return nil
+}
+
+func (c *controller) initDataStore() error {
+	store, err := datastore.NewDataStore(&c.cfg.Datastore)
+	if err != nil {
+		return err
+	}
+	c.store = store
```

Comment:
Please acquire the controller Lock. I know this is only happening in the `controller.New` path right now but since this is a generic function and it may be called in the future in different places it is a good idea